### ISSUE TITLE
[Performance] Linked products section loads all the collection

### ISSUE THREAD

**Issue by tomalowsky** [magento/magento2#21010]
### Preconditions (*)
1. Clean M2.3 installation

### Steps to reproduce (*)
1. Go to the admin panel.
2. Open catalog -> products section.
3. Choose one of the product having a lot of linked children (e. g Related/Up-sells/Cross-sells).
4. Go to edit page of that product.

### Expected result (*)
1. Page loads quickly.

### Actual result (*)
1. Page loads for a long time (or even does not load at all).

Dev note:
I tried above steps for product having more than 5k linked products.
When you get deeper into the core you can see in
` /module-catalog/Ui/DataProvider/Product/Form/Modifier/Related::modifyData($data) `
there for each linked product is loaded model from repository. This approach is not using any pagination, so in case we have a lot of children the whole collection will be loaded at once. 
 It seems the related section is implemented in a wrong way. The product edit page uses ui form component, which needs to collect all product data at once to save it again (with changes applied by user) after "Save" button is clicked. 
My own solution is to move those section to a different controller where ui listing (not form!) will be used. By that, I can use pagination with ajax request. Additional implementation for CRUD operations is also required.

**Comment by magento-engcom-team** [comment]
Hi @tomalowsky. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@tomalowsky do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @GovindaSharma. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by GovindaSharma** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @GovindaSharma. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @GovindaSharma, here is your [Magento instance](https://i-21010-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21010-2-3-develop.instances.magento-community.engineering/admin](https://i-21010-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
@GovindaSharma did you try to generate 5k linked products on vanilla instance ?)

**Comment by GovindaSharma** [comment]
@engcom-backlog-nazar ,i asked instance for another issue as i there was no available instance,so that's why i asked here..so is there any problem if i asked vanilla instance here?

**Comment by ghost** [comment]
@tomalowsky You assigned for this issue so this looks like you working on this.

**Comment by GovindaSharma** [comment]
@engcom-backlog-nazar no as of now i am not working on this may be later i will work,it was just for having vanilla instance,i will remove my assignment shortly.

**Comment by ghost** [comment]
@GovindaSharma ok

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98108`, `MAGETWO-98109` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @tomalowsky. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24145` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
